source/base/mpi.cc

```diff
@@ -812,7 +814,756 @@ namespace Utilities
 #endif
     }
 
+#ifdef DEAL_II_WITH_MPI
+
+#  ifdef DEBUG
+#    define COMPUTE_INDEX_OWNER_USE_OWN_IBARRIER
+#  endif
+
+    namespace ComputeIndexOwner
+    {
+      struct Dictionary
+      {
+        std::vector<unsigned int> array;
+
+        types::global_dof_index dofs_per_process;
+        std::pair<types::global_dof_index, types::global_dof_index> local_range;
+        types::global_dof_index                                     local_size;
+        types::global_dof_index                                     size;
+
+        int offset;
+
+        void
+        reinit(const IndexSet &owned_dofs, const MPI_Comm &comm)
+        {
+#  ifdef DEBUG
+          std::srand(std::time(nullptr));
+          offset = std::rand();
+          offset = ((offset % 32768) / 4) * 4;
+          MPI_Bcast(&offset, 1, MPI_INT, 0, comm);
```

Comment:
I think this block could use a comment :-) I have no idea what you are doing here and why it is randomized.

The use `std::srand` is not thread-safe. Please use the following code:
```
const unsigned int random_seed  = std::random_device()();
std::mt19937 random_number_generator;
random_number_generator.seed(random_seed);
offset = std::uniform_int_distribution<unsigned int> (0, std::numeric_limits<unsigned int>::max())(random_number_generator);
```
or something similar.

Reply:
Or you could just use
```
  offset = std::random_device()();
```